### PR TITLE
Fix build with FreeBSD 1/2

### PR DESCRIPTION
**Pull request by pkubaj**
This is the first patch that fixes build with FreeBSD.

### CODE REVIEW

**Comment by janisozaur** [comment]
I'm a bit confused, the previous code dynamically obtained the executable's path via `KERN_PROC_PATHNAME`, but now it gets degraded to a hardcoded value. Can't you use the previous function call to get the path in runtime anymore?

**Comment by pkubaj** [comment]
Not really.
Compiling this without my patch gives following errors:
```
/usr/local/poudriere/ports/default/games/openrct2/work/OpenRCT2-0.1.2/src/openrct2/platform/Platform.Linux.cpp:150:32: error: use of undeclared identifier 'CTL_KERN'
        const sint32 mib[] = { CTL_KERN, KERN_PROC, KERN_PROC_PATHNAME, -1 };
                               ^
/usr/local/poudriere/ports/default/games/openrct2/work/OpenRCT2-0.1.2/src/openrct2/platform/Platform.Linux.cpp:150:42: error: use of undeclared identifier 'KERN_PROC'
        const sint32 mib[] = { CTL_KERN, KERN_PROC, KERN_PROC_PATHNAME, -1 };
                                         ^
/usr/local/poudriere/ports/default/games/openrct2/work/OpenRCT2-0.1.2/src/openrct2/platform/Platform.Linux.cpp:150:53: error: use of undeclared identifier 'KERN_PROC_PATHNAME'
        const sint32 mib[] = { CTL_KERN, KERN_PROC, KERN_PROC_PATHNAME, -1 };
                                                    ^
/usr/local/poudriere/ports/default/games/openrct2/work/OpenRCT2-0.1.2/src/openrct2/platform/Platform.Linux.cpp:152:13: error: use of undeclared identifier 'sysctl'
        if (sysctl(mib, 4, exePath, &exeLen, nullptr, 0) == -1)
            ^
4 errors generated.
```

Now, those values are defined in `sys/sysctl.h` header. If they are to be used, this header should have been included inside __FreeBSD__ ifdef in the first place. But if I add it myself, the build fails with another error:
```
In file included from /usr/local/poudriere/ports/default/games/openrct2/work/OpenRCT2-0.1.2/src/openrct2/platform/Platform.Linux.cpp:150:
/usr/include/sys/sysctl.h:803:1: error: expected unqualified-id
__BEGIN_DECLS
^
/usr/include/sys/cdefs.h:60:30: note: expanded from macro '__BEGIN_DECLS'
#define __BEGIN_DECLS   extern "C" {
                               ^
/usr/local/poudriere/ports/default/games/openrct2/work/OpenRCT2-0.1.2/src/openrct2/platform/Platform.Linux.cpp:153:13: error: use of undeclared identifier 'sysctl'
        if (sysctl(mib, 4, exePath, &exeLen, nullptr, 0) == -1)
            ^
2 errors generated.
```

I'm not really sure why it fails, it should build just fine. Still, I guess anyone installing OpenRCT2 on FreeBSD does it via ports tree or official binary repository, so binaries land in `/usr/local/bin` anyway.

**Comment by janisozaur** [comment]
Can you link where I can browse those files online?

**Comment by janisozaur** [comment]
I also noticed your other patch targets freebsd 10.4, which is not current. For which version of freebsd is this patch? Does it affect freebsd 11.1 in any way?

**Comment by pkubaj** [comment]
Sure, here are those files:
https://svnweb.freebsd.org/base/head/sys/sys/sysctl.h?view=markup
https://svnweb.freebsd.org/base/head/sys/sys/cdefs.h?view=markup

Keep in mind that this is head, so line numbers may be off with my logs.

My patch targets FreeBSD 10.4, because it's the oldest release still supported (technically, 10.3 is the oldest supported, but it's only a month till EoL and there's not much of a difference between them). But I have tested building OpenRCT2 0.1.2 on FreeBSD 12-CURRENT (with LLVM 6.0.0) and building fails the same way.

**Comment by pkubaj** [comment]
You can have a look at a list of supported versions here:
https://www.freebsd.org/security/index.html#sup

**Comment by janisozaur** [comment]
@pkubaj looks like you tried including that file inside a function. That is not supported use case for this file, please retry with including the file like so:
```diff
diff --git a/src/openrct2/platform/Platform.Linux.cpp b/src/openrct2/platform/Platform.Linux.cpp
index 31f26ec6b..52eacf353 100644
--- a/src/openrct2/platform/Platform.Linux.cpp
+++ b/src/openrct2/platform/Platform.Linux.cpp
@@ -18,6 +18,9 @@
 
 #include <limits.h>
 #include <pwd.h>
+#if defined(__FreeBSD__)
+#include <sys/sysctl.h>
+#endif // __FreeBSD__
 #include "../core/Path.hpp"
 #include "../core/Util.hpp"
 #include "../OpenRCT2.h"
```

**Comment by pkubaj** [comment]
Yes, you were right that I tried to include it in a function. Your patch is fine except that it needs more includes :)
'''
diff --git a/src/openrct2/platform/Platform.Linux.cpp b/src/openrct2/platform/Platform.Linux.cpp
index 31f26ec6b..52eacf353 100644
--- a/src/openrct2/platform/Platform.Linux.cpp
+++ b/src/openrct2/platform/Platform.Linux.cpp
@@ -18,6 +18,9 @@
 
 #include <limits.h>
 #include <pwd.h>
+#if defined(__FreeBSD__)
+#include <stddef.h>
+#include <sys/types.h>
+#include <sys/sysctl.h>
+#endif // __FreeBSD__
 #include "../core/Path.hpp"
 #include "../core/Util.hpp"
 #include "../OpenRCT2.h"
'''

Could you commit it?

**Comment by janisozaur** [comment]
I'd rather take a PR from someone actually using the target platform, can you please update this PR?

File: src/openrct2/platform/Platform.Linux.cpp [hunk 1]
```diff
@@ -150,12 +155,12 @@ namespace Platform
         {
             log_fatal("failed to read /proc/self/exe");
         }
-#elif defined(__FreeBSD__)
-        const sint32 mib[] = { CTL_KERN, KERN_PROC, KERN_PROC_PATHNAME, -1 };
-        auto exeLen = sizeof(exePath);
-        if (sysctl(mib, 4, exePath, &exeLen, nullptr, 0) == -1)
-        {
-            log_fatal("failed to get process path");
+#elif defined(__FreeBSD__)		 +#elif defined(__OpenBSD__) || defined(__FreeBSD__)
```
Comment (janisozaur) [inline src/openrct2/platform/Platform.Linux.cpp]: That doesn't look quite right

Comment (pkubaj) [reply src/openrct2/platform/Platform.Linux.cpp]: Yeah, I noticed it right after sending it.

**Pull request merged**